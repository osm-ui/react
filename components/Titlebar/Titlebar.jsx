import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import classnames from 'classnames';


const StyledHeader = styled.header`
    transition: all 0.1s ease-out;

    left: 0;
    right: 0;
    color: ${props => props.theme.titlebar.color};
    border-color: ${props => props.theme.titlebar.borderColor};
    border-style: ${props => props.theme.titlebar.borderStyle};
    border-width: 0;
    background: ${props => props.theme.titlebar.backgroundColor};
    text-align: center;

    &.container-parent { position: absolute; }
    &.container-root   { position: fixed; }

    &.position-top {
        top: 0;
        border-bottom-width: ${props => props.theme.titlebar.borderWidth};
    }

    &.position-bottom {
        bottom: 0;
        border-top-width: ${props => props.theme.titlebar.borderWidth};
    }

    &.xs {
        height: ${props => props.theme.titlebar.xsHeight};
        font-size: ${props => props.theme.titlebar.xsFontSize};
        line-height: ${props => props.theme.titlebar.xsHeight};
    }

    &.sm {
        height: ${props => props.theme.titlebar.smHeight};
        font-size: ${props => props.theme.titlebar.smFontSize};
        line-height: ${props => props.theme.titlebar.smHeight};
    }

    &.md {
        height: ${props => props.theme.titlebar.mdHeight};
        font-size: ${props => props.theme.titlebar.mdFontSize};
        line-height: ${props => props.theme.titlebar.mdHeight};
    }

    &.lg {
        height: ${props => props.theme.titlebar.lgHeight};
        font-size: ${props => props.theme.titlebar.lgFontSize};
        line-height: ${props => props.theme.titlebar.lgHeight};
    }

    &.maximized {
        height: 100%;
        top: 0;
        bottom: 0;
        border-width: 0;
    }
`;


class Titlebar extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            maximized: props.maximized,
        };
    }

    componentDidMount() {
        if (this.props.maximized === true) {
            this._triggerCallback('onMaximize');
        }
    }

    componentWillReceiveProps(nextProps) {
        this.setState({
            maximized: nextProps.maximized,
        });

        if (this.props.maximized !== nextProps.maximized) {
            if (nextProps.maximized === true) {
                this._triggerCallback('onMaximize');
            }
            else {
                this._triggerCallback('onUnmaximize');
            }
        }
    }

    _triggerCallback(name) {
        if (this.props[name] !== null) {
            this.props[name]();
        }
    }

    render() {
        const {
            position,
            size,
            container,
            className,
            children,
            ...rest
        } = this.props;

        const asideClasses = classnames(className, {
            [size]: true,
            [`position-${position}`]: true,
            [`container-${container}`]: true,
            maximized: this.state.maximized,
        });

        return (
            <StyledHeader className={asideClasses} {...rest}>
                {children}
            </StyledHeader>
        );
    }
}


Titlebar.propTypes = {
    position: PropTypes.oneOf(['top', 'bottom']),
    size: PropTypes.oneOf(['xs', 'sm', 'md', 'lg']),
    container: PropTypes.oneOf(['parent', 'root']),
    maximized: PropTypes.bool,
    onMaximize: PropTypes.func,
    onUnmaximize: PropTypes.func,
    className: PropTypes.string,
    children: PropTypes.node,
};

Titlebar.defaultProps = {
    position: 'top',
    size: 'md',
    container: 'parent',
    maximized: false,
    onMaximize: null,
    onUnmaximize: null,
    className: '',
    children: '',
};

Titlebar.displayName = 'Titlebar';

export default Titlebar;
