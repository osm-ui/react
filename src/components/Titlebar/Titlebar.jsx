import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import classnames from 'classnames';

const StyledHeader = styled.header`
  z-index: 1000;
  transition: all 0.1s ease-out;

  left: 0;
  right: 0;
  font-weight: ${p => p.theme.titlebar.fontWeight};
  color: ${p => p.theme.titlebar.color};
  background: ${p => p.theme.titlebar.backgroundColor};
  box-shadow: ${p => p.theme.titlebar.boxShadow};
  text-align: center;

  &.container-parent {
    position: absolute;
  }
  &.container-root {
    position: fixed;
  }

  &.position-top {
    top: 0;
    border-bottom-width: ${p => p.theme.titlebar.borderWidth};
  }

  &.position-bottom {
    bottom: 0;
    border-top-width: ${p => p.theme.titlebar.borderWidth};
  }

  &.xs {
    height: ${p => p.theme.titlebar.xsHeight};
    font-size: ${p => p.theme.titlebar.xsFontSize};
    line-height: ${p => p.theme.titlebar.xsHeight};
  }

  &.sm {
    height: ${p => p.theme.titlebar.smHeight};
    font-size: ${p => p.theme.titlebar.smFontSize};
    line-height: ${p => p.theme.titlebar.smHeight};
  }

  &.md {
    height: ${p => p.theme.titlebar.mdHeight};
    font-size: ${p => p.theme.titlebar.mdFontSize};
    line-height: ${p => p.theme.titlebar.mdHeight};
  }

  &.lg {
    height: ${p => p.theme.titlebar.lgHeight};
    font-size: ${p => p.theme.titlebar.lgFontSize};
    line-height: ${p => p.theme.titlebar.lgHeight};
  }

  &.maximized {
    height: 100%;
    top: 0;
    bottom: 0;
    border-width: 0;
  }
`;

class Titlebar extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      maximized: props.maximized
    };
  }

  componentDidMount() {
    if (this.props.maximized === true) {
      this._triggerCallback('onMaximize');
    }
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      maximized: nextProps.maximized
    });

    if (this.props.maximized !== nextProps.maximized) {
      if (nextProps.maximized === true) {
        this._triggerCallback('onMaximize');
      } else {
        this._triggerCallback('onUnmaximize');
      }
    }
  }

  _triggerCallback(name) {
    if (this.props[name] !== null) {
      this.props[name]();
    }
  }

  render() {
    const {
      position,
      size,
      container,
      className,
      children,
      ...rest
    } = this.props;

    const asideClasses = classnames(className, {
      [size]: true,
      [`position-${position}`]: true,
      [`container-${container}`]: true,
      maximized: this.state.maximized
    });

    return (
      <StyledHeader className={asideClasses} {...rest}>
        {children}
      </StyledHeader>
    );
  }
}

Titlebar.propTypes = {
  position: PropTypes.oneOf(['top', 'bottom']),
  size: PropTypes.oneOf(['xs', 'sm', 'md', 'lg']),
  container: PropTypes.oneOf(['parent', 'root']),
  maximized: PropTypes.bool,
  onMaximize: PropTypes.func,
  onUnmaximize: PropTypes.func,
  className: PropTypes.string,
  children: PropTypes.node
};

Titlebar.defaultProps = {
  position: 'top',
  size: 'md',
  container: 'parent',
  maximized: false,
  onMaximize: null,
  onUnmaximize: null,
  className: '',
  children: ''
};

Titlebar.displayName = 'Titlebar';

export default Titlebar;
