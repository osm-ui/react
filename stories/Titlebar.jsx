import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { host } from 'storybook-host';
import { withKnobs, boolean, select } from '@storybook/addon-knobs';

import defaultHostOptions from './defaultHostOptions';
import KnobsAlert from './components/KnobsAlert';
import FakeApp from './components/FakeApp';

import {
  DefaultTheme,
  WhiteTheme,
  LightGrayTheme,
  DarkGrayTheme,
  AnthraciteTheme,
  YellowTheme,
  OrangeTheme,
  BrownTheme,
  RedTheme,
  RoseTheme,
  PurpleTheme,
  BlueTheme,
  SkyTheme,
  TurquoiseTheme,
  GreenTheme,
  Titlebar
} from 'index';

const fakeAppStyle = {
  paddingTop: '70px',
  paddingBottom: '70px'
};

storiesOf('Titlebar', module)
  .addDecorator(withKnobs)
  .addDecorator(
    host({
      ...defaultHostOptions,
      title: 'Titlebar'
    })
  )
  .addWithInfo('Default state', () => (
    <DefaultTheme>
      <FakeApp style={fakeAppStyle}>
        <Titlebar>Application title</Titlebar>
      </FakeApp>
    </DefaultTheme>
  ))
  .addWithInfo('Position', () => (
    <DefaultTheme>
      <FakeApp style={fakeAppStyle}>
        <Titlebar position="top">Application header</Titlebar>
        <Titlebar position="bottom">Application footer</Titlebar>
      </FakeApp>
    </DefaultTheme>
  ))
  .addWithInfo('Theme', () => (
    <DefaultTheme>
      <FakeApp style={fakeAppStyle}>
        <BlueTheme>
          <Titlebar>Application header</Titlebar>
        </BlueTheme>
        <OrangeTheme>
          <Titlebar position="bottom">Application footer</Titlebar>
        </OrangeTheme>
      </FakeApp>
    </DefaultTheme>
  ))
  .addWithInfo('Size', () => (
    <DefaultTheme>
      <FakeApp style={fakeAppStyle}>
        <BlueTheme>
          <Titlebar size="lg">Application header</Titlebar>
          <Titlebar size="sm" position="bottom">
            Application footer
          </Titlebar>
        </BlueTheme>
      </FakeApp>
    </DefaultTheme>
  ))
  .addWithInfo('Playground', () => {
    const position = select('Position', ['top', 'bottom'], 'top');
    const size = select('Size', ['xs', 'sm', 'md', 'lg'], 'md');
    const container = select('Container', ['parent', 'root'], 'parent');
    const maximized = boolean('Maximized', false);
    const theme = select(
      'Theme',
      [
        'Default',
        'White',
        'LightGray',
        'DarkGray',
        'Anthracite',
        'Yellow',
        'Orange',
        'Brown',
        'Red',
        'Rose',
        'Purple',
        'Blue',
        'Sky',
        'Turquoise',
        'Green'
      ],
      'Rose'
    );
    const themes = {
      DefaultTheme,
      WhiteTheme,
      LightGrayTheme,
      DarkGrayTheme,
      AnthraciteTheme,
      YellowTheme,
      OrangeTheme,
      BrownTheme,
      RedTheme,
      RoseTheme,
      PurpleTheme,
      BlueTheme,
      SkyTheme,
      TurquoiseTheme,
      GreenTheme
    };
    const ThemeElement = themes[`${theme}Theme`];

    return (
      <DefaultTheme>
        <KnobsAlert />
        <FakeApp style={fakeAppStyle}>
          <ThemeElement>
            <Titlebar
              maximized={maximized}
              position={position}
              container={container}
              size={size}
              onMaximize={action('onMaximize titlebar')}
              onUnmaximize={action('onUnmaximize titlebar')}
            >
              Application title
            </Titlebar>
          </ThemeElement>
        </FakeApp>
      </DefaultTheme>
    );
  });
