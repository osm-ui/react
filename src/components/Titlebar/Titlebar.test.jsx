import React from 'react';
import {
  snapshotWithElementChildren,
  snapshotWithTextChildren
} from '../../helpers/tests';
import Titlebar from './';

describe('When using snapshots', () => {
  it('Should render with an element children', () =>
    snapshotWithElementChildren(Titlebar));

  it('Should render with a text children', () =>
    snapshotWithTextChildren(Titlebar));
});

describe('When using snapshots with the maximized prop', () => {
  it('Should render with an element children', () =>
    snapshotWithElementChildren(Titlebar, { maximized: true }));

  it('Should render with a text children', () =>
    snapshotWithTextChildren(Titlebar, { maximized: true }));
});
