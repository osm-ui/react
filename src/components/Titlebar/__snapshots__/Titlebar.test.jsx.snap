// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When using snapshots Should render with a text children 1`] = `
<span
  className="sc-bdVaJa lbJuWU"
>
  <header
    className="md position-top container-parent sc-bwzfXH Wioxr"
  >
    This is a text
  </header>
</span>
`;

exports[`When using snapshots Should render with an element children 1`] = `
<span
  className="sc-bdVaJa lbJuWU"
>
  <header
    className="md position-top container-parent sc-bwzfXH Wioxr"
  >
    <div />
  </header>
</span>
`;

exports[`When using snapshots with the maximized prop Should render with a text children 1`] = `
<span
  className="sc-bdVaJa lbJuWU"
>
  <header
    className="md position-top container-parent maximized sc-bwzfXH Wioxr"
  >
    This is a text
  </header>
</span>
`;

exports[`When using snapshots with the maximized prop Should render with an element children 1`] = `
<span
  className="sc-bdVaJa lbJuWU"
>
  <header
    className="md position-top container-parent maximized sc-bwzfXH Wioxr"
  >
    <div />
  </header>
</span>
`;
